### PR TITLE
YAFA

### PR DESCRIPTION
**Pull request by gradha**
Yet another _fine_ applicant.

I was meaning to start today and do the exercise over the weekend but the recruiter told me yesterday I only had until today, so hopefully this also makes it for evaluation. In total this work clocks around 14h 30m, which fits the requirement of 3 days if you count 8h per day.

The project is not complete according to the tasks [I was creating for myself](https://github.com/gradha/theranking-ios-candidates/issues) but should be enough to evaluate code quality as it meets (I hope) the minimum requirements. Many more enhancements could be made, but I understand that's not the point and the schedule is tight.

### CODE REVIEW

**Comment by caldofran** [comment]
Hi Grzegorz, first of all, thanks for spending your time making this test. In general, the test is good, but I would like to make some points clearer.
I understand due to the close schedule you may take less care of some details, but I would like to know the way you think while developing or some decisions you have taken.

For instance, I realize you use very often MACROS for some tasks. I would like you to explain me a little bit the reason why you prefer MACROS instead of other possible solutions.

Another thing that worries me a bit is that some methods have very long implementations. For me, it was a little difficult to follow the implementation of some parts of your code.

To sum up, I like your test and I would like to talk with your about some development details deeply.

Thanks for your effort.

**Comment by gradha** [comment]
I favour macro usage for conditional compilation like the [DLOG](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/global/ELHASO.h#L21) or [LASSERT](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/global/ELHASO.h#L54) macros which are meant to go away on release builds. The usage of constants like [kConsumerKey](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/global/TRSecrets.h#L4) allows a more elaborate `#ifdef` compilation check which prevents people from compiling source code unless they fill in important files out source control. It allows to raise early human friendly compile errors.

I also like them for reducing code verbosity like the [_ADD](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/models/TRPhotoData.m#L135) macro because it simplifies code. The prefix underscore signals this is a very local `#define` and typically I'd write after its usage `#undef _ADD` to prevent polluting further code, but I forgot.

The remaining _suspicious_ usage would be constants like [kBaseHost](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/global/TRLogic.m#L11), [kErrorLogic](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/global/TRLogic.h#L4) or [kTRGridPhotoCell](https://github.com/gradha/theranking-ios-candidates/blob/develop/src/vcs/TRGridVC.m#L12). In general the accepted pattern is to write these as `extern NSString*`, which I tend to not do because I only find one good reason for this.

One argument is that an `extern NSString*` gives the compiler more typing information to catch missuses. This doesn't make much sense in a weakly typed language like Objective-C, plus you can't write `someString + someInteger` and have it compile due to lack of operator overloading. But the non constness prevents you from using the symbol with the preprocessor.

Another argument is that if you export a `#define` symbol in a header you later might want to change it. Duh, it's a constant, if you are changing it you are breaking code anyway and should provide a documented depreaction path anyway. However symbol exporting is to me the real reason one would use an `extern NSString*`, because [you can check at runtime if a global symbol exists](http://stackoverflow.com/a/11041798/172690) and this also plays nicer with libaries.

That's precisely the reason I _prefer_ to use macros, because those symbols are **not** for exported/public API usage, yet sometimes they creep into header files. Having them as `#define` is a warning to not rely on them, in a similar fashion to header comments saying `don't use this because it is a private API`. The `kErrorLogic` symbol is a mistake though. While in practice nothing wrong will happen because I'm not writing a public use API, header symbols should at least try to follow Objective-C/Cocoa conventions and this one looks odd.

To avoid `#defines` creeping into public headers as exported symbols I tend to use [a very different naming convention](https://github.com/gradha/Record-my-position/blob/master/src/GPS.m#L7) using upper case and prefix underscore so that it is very visible the constant is not a public API. But this convention is unusual (like the snake_case convention) and I didn't want to scare so I reused the `kSomething` pattern.

With regards to the long implementation of some methods I totally agree. I don't get much chance to use newer APIs at work (just this year was the first time I was able to use ARC in a project) and forgot to mention that I wanted to experiment with `NSURLSession` which I used for the first time. So far all of my network code has used delegates and kept state between calls with instance variables, but I wanted to give it a go and try the _blocks are for hipsters_ API. Which resulted in the long `fetchPhotosWithCallback`.

Not very representative of typical network code I write, which I tend to push off into a separate network class. Since this example has a single network request API there's not much point in splitting the long sequence of parsing and error checking into separate methods, but certainly once you add a few more requests it is better to factor some of it to make it more manageable and avoid repetition. Plus in a real world app you would like to keep track of requests to cancel them or at least make the internal API more resilient to reentrancy in case of missuse.
